src/core/namespace.c

```diff
@@ -3310,6 +3312,373 @@ bool ns_type_supported(NamespaceType type) {
+         *    access to their FDs
+         * 2. Fork into a grandchild, which will enter the target namespace and attempt to "unpeel" the
+         *    overlays to obtain FDs the underlying directories, over which we will reapply the overlays
+         * 3. In the child again, receive the FDs and reapply the overlays
+         */
+        r = safe_fork("(sd-ns-refresh-exts)",
+                        FORK_DEATHSIG_SIGTERM|FORK_WAIT|FORK_NEW_MOUNTNS|FORK_MOUNTNS_SLAVE,
+                        NULL);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                /* Child (host namespace) */
+                _cleanup_close_ int mntns_fd = -EBADF, root_fd = -EBADF, pidns_fd = -EBADF;
+                _cleanup_close_pair_ int pair[2] = EBADF_PAIR;
+                _cleanup_(sigkill_waitp) pid_t grandchild_pid = 0;
+
+                 (void) mkdir_p_label(overlay_prefix, 0555);
+
+                /* Open all extension roots on the host */
+                FOREACH_ARRAY(m, ml.mounts, ml.n_mounts)
+                        if (IN_SET(m->mode, MOUNT_EXTENSION_DIRECTORY, MOUNT_EXTENSION_IMAGE)) {
+                                r = apply_one_mount("/", m, p);
+                                if (r < 0) {
+                                        log_debug_errno(r, "Failed to apply extension mount: %m");
+                                        _exit(EXIT_FAILURE);
+                                }
+                        }
+
+                /* Create a grandchild process to handle the unmounting and reopening of hierarchy */
+                r = socketpair(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0, pair);
+                if (r < 0) {
+                        log_debug_errno(errno, "Failed to create socket pair: %m");
+                        _exit(EXIT_FAILURE);
+                }
+
+                r = safe_fork("(sd-ns-refresh-exts-grandchild)",
+                                FORK_LOG|FORK_DEATHSIG_SIGKILL,
+                                &grandchild_pid);
+                if (r < 0)
+                        _exit(EXIT_FAILURE);
+                if (r == 0) {
+                        /* Grandchild (target service namespace) */
+                        pair[0] = safe_close(pair[0]);
+
+                        r = namespace_open(target->pid, &pidns_fd, &mntns_fd, NULL, NULL, &root_fd);
+                        if (r < 0) {
+                                log_debug_errno(r, "Failed to open namespace: %m");
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        r = namespace_enter(pidns_fd, mntns_fd, -EBADF, -EBADF, root_fd);
+                        if (r < 0) {
+                                log_debug_errno(r, "Failed to enter namespace: %m");
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        /* Prevent propagation while unmounting */
+                        r = mount(NULL, "/", NULL, MS_SLAVE|MS_REC, NULL);
```

Comment:
error handling?

Reply:
i don't get this. why is this necessary?

Reply:
isn't this needed to avoid the umount from affecting the running process? We are doing this in a copy to prep the new overlayfs, but it shouldn't be visibile to the process until the very last step

Reply:
this is done in the namespace of the "target", we enter that namespace immediately before. hence, this doesn't seem right to me?

Reply:
Yes this looks like a leftover, now there's a child process in a cloned namespace